Add Objection w/ initial schema and seeds, and auth w/ crypto

SamHH opened the pull request:
This PR includes database connectivity, initial schema, initial seeds, and the automatic creation of a new player/account on server start as a demo.

Included also is MJS so that we can use modern JS import/export syntax isomorphically.

RonSpawnson commented on server/db/migrations/20171020144702_initial_schema.js:
```diff
@@ -0,0 +1,41 @@
+			table.integer('selected_ball').references('id').inTable('ball').onDelete('SET NULL')
+			table.specificType('competitive_honors', 'text[]')
+		})
+		.createTable('map', table => {
```
Perhaps maps should have author information. Also note, some maps have more than 1 author

samhh replied:
Good shout, I'll add an authors key as something like `varchar(n)[]`.

Worth adding the same to flairs and balls or no?

sallaben replied:
Since when do flairs and balls have authors? Or are you implying something else?

samhh replied:
I'm imagining a system whereby flairs/balls/other assets are contributed to the game by the community. In that case they may like accreditation as map authors do.

SamHH replied:
How do we want authors to be denoted? I was thinking an array of strings, but then I wondered if we might prefer actually linking them to the players i.e. an array of player IDs (corresponding to the player table) instead.

whackashoe commented on server/db/migrations/20171020144702_initial_schema.js:
```diff
@@ -0,0 +1,41 @@
+			table.string('description', 100)
+			table.text('filename').notNullable()
+		})
+		.createTable('ball', table => {
```
what is the point of this table? for custom balls?

samhh replied:
Yep, that's the idea. That can be removed for the time being though as it isn't fleshed out and certainly won't be in MVP.

whackashoe commented on server/db/migrations/20171020144702_initial_schema.js:
```diff
@@ -0,0 +1,41 @@
+			table.string('description', 100)
+			table.text('filename').notNullable()
+		})
+		.createTable('player', table => {
```
wins, losses?

samhh replied:
Yes... this needs discussion I think about what stats we're storing and in what format.

whackashoe commented on server/db/migrations/20171020144702_initial_schema.js:
```diff
@@ -0,0 +1,41 @@
+exports.up = knex =>
```
would be cool to have games table which stores all existing game stats,

players
joined_at

who scored
who won
etc

samhh replied:
Would this be for temporary use? If so, probably worth consulting with @OpenSourceCTF/engine requirements.

sallaben commented:
That's a good idea, actually. Maybe we could create an in-browser flair
creator. That would put a spin on things.
On Fri, Nov 3, 2017 at 7:37 AM Sam A. Horvath-Hunt <notifications@github.com>
wrote:

> *@SamHH* commented on this pull request.
> ------------------------------
>
> In server/db/migrations/20171020144702_initial_schema.js
> <https://github.com/OpenSourceCTF/website/pull/17#discussion_r148761001>:
>
> > +			table.string('description', 100)
> +			table.text('filename').notNullable()
> +		})
> +		.createTable('player', table => {
> +			table.uuid('id').primary().defaultTo(knex.raw('uuid_generate_v4()')).notNullable()
> +			table.dateTime('created_at').defaultTo(knex.fn.now()).notNullable()
> +			table.dateTime('last_active').defaultTo(knex.fn.now()).notNullable()
> +			table.string('username', 20).unique().notNullable()
> +			table.dateTime('last_modified_username')
> +			table.string('email', 254).unique()
> +			table.string('password', 100).notNullable() // TODO update this length to reflect pw crypto
> +			table.integer('selected_flair').references('id').inTable('flair').onDelete('SET NULL')
> +			table.integer('selected_ball').references('id').inTable('ball').onDelete('SET NULL')
> +			table.specificType('competitive_honors', 'text[]')
> +		})
> +		.createTable('map', table => {
>
> I'm imagining a system whereby flairs/balls/other assets are contributed
> to the game by the community. In that case they may like accreditation as
> map authors do.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenSourceCTF/website/pull/17#discussion_r148761001>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHV7IyxiXbWfXNtq85vEMPMzEZ3fZb7sks5syvsVgaJpZM4QKdGJ>
> .
>

SamHH commented:
I've removed the balls and flairs tables as we probably won't be touching this for MVP (and if we are, we can easily add them back in with better requirements). Same goes for competitive honours column.

I've added an authors column to the map table as an array of strings.

What's left for this PR:

- Determine what stats we want to store, how we want to store them, and whether that's something for MVP or not.

sallaben commented:
When we do implement it, I think stats should be stored by the engine
during each game and committed to the database whenever the user
disconnects from that game (whether through quitting or joining a new one
when that game ends). What do you think?
On Tue, Nov 7, 2017 at 6:42 AM Sam A. Horvath-Hunt <notifications@github.com>
wrote:

> I've removed the balls and flairs tables as we probably won't be touching
> this for MVP (and if we are, we can easily add them back in with better
> requirements). Same goes for competitive honours column.
>
> I've added an authors column to the map table as an array of strings.
>
> What's left for this PR:
>
>    - Determine what stats we want to store, how we want to store them,
>    and whether that's something for MVP or not.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenSourceCTF/website/pull/17#issuecomment-342457724>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHV7I6_k3V8IZB4CN6R7VbrvT_F5F668ks5s0EIdgaJpZM4QKdGJ>
> .
>

SamHH commented:
Note to self: All that's left before this branch can be merged is actioning Contessa's feedback in Discord/#backend.

SamHH commented:
RE: the above, that's all that needs doing now. The other advice was regards matches which can come later. Time this PR was merged!
